The job **`x86_64-gnu-distcheck`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13101999054/job/36551411568) [(plain)](https://github.com/rust-lang-ci/rust/commit/20e70db0738cd263b90e8a2bbd01d594bf09eeb0/checks/36551411568/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
[RUSTC-TIMING] volatile_fat_ptr test:true 0.238
[RUSTC-TIMING] run_time_detect test:true 0.478
[RUSTC-TIMING] log_knows_the_names_of_variants_in_std test:true 0.370
[RUSTC-TIMING] test test:true 15.970
rustc: /checkout/src/llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = llvm::PossiblyDisjointInst; From = llvm::Value]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
[RUSTC-TIMING] allocbenches test:true 67.653
[RUSTC-TIMING] coretests test:true 71.221
[RUSTC-TIMING] coretests test:true 71.221
rustc exited with signal: 6 (SIGABRT) (core dumped)
error: could not compile `coretests` (test "coretests")
Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/tmp/distcheck/build/bootstrap/debug/rustc /checkout/obj/build/tmp/distcheck/build/bootstrap/debug/rustc --crate-name coretests --edition=2021 library/coretests/tests/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --emit=dep-info,link -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on --test --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values())' -C metadata=250ed1c368617f52 -C extra-filename=-53348fcd566045d1 --out-dir /checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage1-std/release/deps --extern coretests=/checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/libcoretests-4048b5149784de40.rlib --extern rand=/checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librand-b29c5ee1f82c350f.rlib --extern rand_xorshift=/checkout/obj/build/tmp/distcheck/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/librand_xorshift-2236a71372225502.rlib -Csymbol-mangling-version=legacy '--check-cfg=cfg(feature,values(any()))' -Zunstable-options '--check-cfg=cfg(bootstrap)' -Zmacro-backtrace -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Zinline-mir-preserve-debug -Zmir_strip_debuginfo=locals-in-tiny-functions -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Alinker-messages -Cembed-bitcode=yes -Cforce-frame-pointers=yes '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' -Z binary-dep-depinfo` (exit status: 254)
[RUSTC-TIMING] alloc test:true 103.817
[RUSTC-TIMING] alloctests test:true 115.596
[RUSTC-TIMING] std test:true 142.687
Build completed unsuccessfully in 0:44:34
Build completed unsuccessfully in 0:44:34
make: *** [Makefile:49: check] Error 1
Command has failed. Rerun with -v to see more details.
  local time: Sun Feb  2 20:56:09 UTC 2025
  network time: Sun, 02 Feb 2025 20:56:09 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_codegen_llvm/src/builder.rs
+++ b/compiler/rustc_codegen_llvm/src/builder.rs
@@ -424,7 +424,13 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
     fn or_disjoint(&mut self, a: &'ll Value, b: &'ll Value) -> &'ll Value {
         unsafe {
             let or = llvm::LLVMBuildOr(self.llbuilder, a, b, UNNAMED);
-            llvm::LLVMSetIsDisjoint(or, True);
+
+            // If a and b are both values, then `or` is a value, rather than
+            // an instruction, so we need to check before setting the flag.
+            // (See also `LLVMBuildNUWNeg` which also needs a check.)
+            if llvm::LLVMIsAInstruction(or).is_some() {
+                llvm::LLVMSetIsDisjoint(or, True);
+            }
             or
         }
     }

--- a/tests/codegen/intrinsics/disjoint_bitor.rs
+++ b/tests/codegen/intrinsics/disjoint_bitor.rs
@@ -1,4 +1,4 @@
-//@ compile-flags: -C no-prepopulate-passes
+//@ compile-flags: -C no-prepopulate-passes -Z mir-opt-level=0
 
 #![crate_type = "lib"]
 #![feature(core_intrinsics)]
@@ -17,4 +17,14 @@ pub unsafe fn disjoint_bitor_signed(x: i32, y: i32) -> i32 {
 pub unsafe fn disjoint_bitor_unsigned(x: u64, y: u64) -> u64 {
     // CHECK: or disjoint i64 %x, %y
     disjoint_bitor(x, y)
+}
+
+// CHECK-LABEL: @disjoint_bitor_literal
+#[no_mangle]
+pub unsafe fn disjoint_bitor_literal() -> u8 {
+    // This is a separate check because even without any passes,
+    // LLVM will fold so it's not an instruction, which can assert in LLVM.
+
+    // CHECK: store i8 3
+    disjoint_bitor(1, 2)
 }

